### DISCUSSION TITLE
How to restrict the rod's DOF in a determined position in coordinate system?

### DISCUSSION THREAD

**Discussion by Edmund-Wang0546** [GazzolaLab/PyElastica#198] [Q&A]
How to restrict the rod's DOF in a determined position in coordinate system? Like: 
![2](https://user-images.githubusercontent.com/113122422/192674358-50b9cf91-e241-47b7-a911-a6c75fa3b603.png)
How to set the Constraints of the rod to make it realize the restricted DOF only with translation and rotation in Z (along the hole)? The ' GeneralConstraint' in 'Constraints' could only restrict the rod's motion by position-indices of the rod but not a determined position in the space.

**Comment by armantekinalp** [comment]
Yes purpose of `GeneralConstrain` is to prevent rod to move or rotate based on the initial condition. Can you clarify what is your initial condition? Are the middle elements of the rod initialized inside the hole?

**Reply by Edmund-Wang0546** [reply]
Thanks for your reply. 
Are the middle elements of the rod initialized inside the hole?——Yes! 
And the red board in the figure with hole is rigid body. But PyElastica only provide 'cylinder' and 'sphere' class. Thus, it's impossible to built a rigid body with hole and add 'contact' between hole and rod to constrain the motion of rod. According to this, I plan to restrict one or several DOF of the rod's elements which located in the hole. This means the hold's position is fixed. The restricted element(s) on the rod is changing with the time.

**Reply by armantekinalp** [reply]
I see so the hole position is fixed in space and elements inside the whole can only move up and down (lets says x direction and can rotate). It seems to me you need a custom boundary condition, in which if a rod element enters the hole then constrain y, z translation and 2 rotations. Currently PyElastica does not have this boundary conditions, so you need to implement. If you are interested in we can continue discussing in this thread.
